src/_pytest/mark/structures.py

```diff
@@ -200,21 +201,38 @@ def _for_parametrize(
 
 
 @final
-@attr.s(frozen=True)
+@attr.s(frozen=True, init=False, auto_attribs=True)
 class Mark:
     #: Name of the mark.
-    name = attr.ib(type=str)
+    name: str
     #: Positional arguments of the mark decorator.
-    args = attr.ib(type=Tuple[Any, ...])
+    args: Tuple[Any, ...]
     #: Keyword arguments of the mark decorator.
-    kwargs = attr.ib(type=Mapping[str, Any])
+    kwargs: Mapping[str, Any]
 
     #: Source Mark for ids with parametrize Marks.
-    _param_ids_from = attr.ib(type=Optional["Mark"], default=None, repr=False)
+    _param_ids_from: Optional["Mark"] = attr.ib(default=None, repr=False)
     #: Resolved/generated ids with parametrize Marks.
-    _param_ids_generated = attr.ib(
-        type=Optional[Sequence[str]], default=None, repr=False
-    )
+    _param_ids_generated: Optional[Sequence[str]] = attr.ib(default=None, repr=False)
```

Comment:
should we introduce a metaclass that makgles the `_ispytest` in `__call__` - making all those classes have strane inits is a pain

Reply:
IMHO I prefer the `_is_pytest` approach because it is more explicit... I think a metaclass introduces needless complexity just for this change.

Just stating my opinion though, not against changing it to a metaclass if Ran also prefers to do so.

Reply:
I feel simillary to @nicoddemus, I think that while the `_ispytest` is a bit annoying, it is better in most other dimensions (like ease of understanding, greppability, ease of type checking, ease of bypassing when needed), so I prefer it.